socketioxide/src/payload_value/mod.rs

```diff
@@ -0,0 +1,490 @@
+    }
+}
+
+fn sort_bin_payloads<T>(mut bins: Vec<(usize, T)>) -> Vec<T> {
+    bins.sort_by(|(a, _), (b, _)| a.cmp(b));
+    bins.into_iter().map(|(_, bin)| bin).collect()
+}
+
+mod serde_payload_value_binary {
+    use std::fmt;
+
+    use serde::{de, ser::SerializeMap, Deserializer, Serializer};
+    use serde_json::Value;
+
+    pub fn serialize<S>(num: &usize, _data: &[u8], serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut map = serializer.serialize_map(Some(2))?;
+        map.serialize_entry(&"_placeholder", &true)?;
+        map.serialize_entry(&"num", num)?;
+        map.end()
+    }
+
+    pub fn deserialize<'de, D>(deserializer: D) -> Result<(usize, Vec<u8>), D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct BinaryVisitor;
+
+        impl<'de> de::Visitor<'de> for BinaryVisitor {
+            type Value = (usize, Vec<u8>);
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("a _placeholder object with a num field")
+            }
+
+            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+            where
+                A: de::MapAccess<'de>,
+            {
+                use de::Error;
+
+                let (first_key, first_value) = map
+                    .next_entry::<String, serde_json::Value>()?
+                    .ok_or(A::Error::custom(
+                        "not enough values in object for pending binary",
+                    ))?;
+                let (second_key, second_value) = map
+                    .next_entry::<String, serde_json::Value>()?
+                    .ok_or(A::Error::custom(
+                        "not enough values in object for pending binary",
+                    ))?;
+
+                let placeholder_value = if first_key == "_placeholder" {
+                    Ok(&first_value)
+                } else if second_key == "_placeholder" {
+                    Ok(&second_value)
+                } else {
+                    Err(A::Error::custom("no '_placeholder' present"))
+                }?;
+
+                let num_value = if first_key == "num" {
+                    Ok(&first_value)
+                } else if second_key == "num" {
+                    Ok(&second_value)
+                } else {
+                    Err(A::Error::custom("no 'num' present"))
+                }?;
+
+                match (placeholder_value, num_value) {
+                    (Value::Bool(true), Value::Number(num))
+                        if num.as_u64().filter(|n| *n <= usize::MAX as u64).is_some() =>
+                    {
+                        Ok((num.as_u64().unwrap() as usize, vec![]))
+                    }
+                    _ => Err(A::Error::custom(
+                        "_placeholder & num are of incorrect types or values",
+                    )),
+                }
+            }
+        }
```

Comment:
I would refactor that to:
```rust
fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
        where
            A: de::MapAccess<'de>,
        {
            use de::Error;

            // expect { "_placeholder": true, num: usize }
            for _ in 0..2 {
                let key = map.next_key::<&str>()?.ok_or(A::Error::custom(
                    "the _placeholder object must have exactly two fields",
                ))?;
                match key {
                    "_placeholder" if map.next_value::<bool>()? => {}
                    "_placeholder" => {
                        Err(A::Error::custom("expected true for _placeholder"))?;
                    }
                    "num" => {
                        let num = map.next_value::<usize>()?;
                        return Ok((num, vec![]));
                    }
                    _ => return Err(A::Error::custom("unexpected key in _placeholder object")),
                }
            }

            Err(A::Error::custom(
                "could not find _placeholder and num fields in _placeholder object",
            ))
        }
```

And maybe flatten the impl rather than putting it inside the `deserialize` fn.

Reply:
Ah yeah, your version is much better :)

As it turns out, though, I ended up deleting that deserializer entirely and approached the problem a different way...